docs/md/guides: database per tenant (#2990)
ariga/atlas: doc/md/guides/database-per-tenant/01-intro.mdx

---
title: Database-per-Tenant Architectures with Atlas
id: intro
slug: intro
---

## Intro

This guide describes how to utilize Atlas to manage database schemas in "Database per Tenant" architectures,
a common pattern for deploying multi-tenant applications.

### What is a Database per Tenant Architecture?

> In a "Database per Tenant" architecture, each tenant has its own dedicated database. Database-per-tenant architectures are commonly

Reviewer: ```suggestion
In a "Database per Tenant" architecture, each tenant has its own dedicated database (or a schema). Database-per-tenant architectures are commonly
```

schema, and compute resources.

### Advantages

Database per Tenant architectures offer several advantages over other multi-tenant architectures:

1. **Data Isolation**: Each tenant has its own dedicated database, providing strong data isolation between tenants.
By isolating data on the database level, you can ensure that data from one tenant is not accessible to another tenant,
providing better security and privacy guarantees.
2. **Data Compliance**: By isolating tenants on the database level, you can ensure that each tenant's data is stored in
compliance with data regulations and privacy laws relevant to that tenant. For example, you can host the specific
database in a region that complies with the tenant's data residency requirements or even in your customer's own
infrastructure.
3. **Performance**: By isolating tenants on the database level, you can ensure that the performance of one tenant's
database operations does not impact the performance of other tenants - eliminating the "noisy neighbor" problem.
Additionally, you can utilize different instance types or configurations for each tenant's database to meet their
specific performance requirements.

> 4. **Simplified Development** (compared to shared database architectures) Codebases for multi-tenant applications are

Reviewer 2: ```suggestion
4. **Simplified Development** (compared to shared database architectures): Codebases for multi-tenant applications are
```

notoriously complex. Every action against the database must be made "tenant-aware" to ensure that data is correctly
isolated. By isolating tenants on the database level, applications become effectively single-tenant, simplifying
development, reducing the mental overhead of handling multi-tenancy, and reducing the risk of data leakage between tenants.

> 5. **Simplified Deployment** (compared to full isolation) In a full isolation architecture, each tenant has its own

Reviewer 2: ```suggestion
5. **Simplified Deployment** (compared to full isolation): In a full isolation architecture, each tenant has its own
```

deployment, requiring separate infrastructure, monitoring, and maintenance. By isolating tenants on the database level,
you can share infrastructure, monitoring, and maintenance across tenants, reducing operational overhead.

> 

Reviewer: ```suggestion
6. **Simplified Maintenance**: Running tenants on different databases simplifies maintenance tasks: Deleting, exporting, and loading tenant data do not require special toolings as most databases offer these functionalities.  
```

### Challenges

> > This architecture decisions is one of my biggest regrets, and we are currently in the process of rebuilding into a single database model.

Reviewer 2: ```suggestion
> This architecture decision is one of my biggest regrets, and we are currently in the process of rebuilding into a single database model.
```

Author: this is a quote.

>
> *[HN Thread on Database per Tenant Architecture](https://news.ycombinator.com/item?id=23305111)*

Despite their numerous advantages, Database per Tenant architectures present unique challenges, mostly around managing
database schema migrations:

* **Migration Duration** - As the number of tenants grows, deployment time increases linearly with the number of tenants.
This is especially problematic when deploying schema migrations that take time, such as data migrations or index rebuilds.

> * **Schema Consistency** Ensuring that all tenant databases have consistent schema versions can be complex. Any schema

Reviewer 2: ```suggestion
* **Schema Consistency** - Ensuring that all tenant databases have consistent schema versions can be complex. Any schema
```

change needs to be applied across all tenant databases, which can be time-consuming and error-prone.
* **Detecting Drift** - Migrations can fail due to various reasons, such as network issues, database downtime,

> or data dependent issues (such as constraint violations).  Such inconsistencies can cause application downtime for

Reviewer 2: ```suggestion
or data dependent issues (such as constraint violations). Such inconsistencies can cause application downtime for
```

users, as new application versions are rolled out and expect database changes to already have been applied.
Detecting these inconsistencies (schema-drift) across a large fleet of databases can be challenging.

> * **Rollbacks** - Rolling back schema changes across a large fleet of databases can be challenging.  In a shared database

Reviewer 2: ```suggestion
* **Rollbacks** - Rolling back schema changes across multiple databases can be difficult. In a shared database
```
You use the same sentence one above

architecture, rolling back a schema change is a single operation. In a database per tenant architecture, rolling back

> a schema change requires rolling back the change on each tenant's database, which needs to be somehow orchestrated.

Reviewer 2: ```suggestion
a schema change requires rolling back the change on each tenant's database, which needs to be orchestrated somehow.
```
